FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/DataFilter/ContainerRegistryFilterProvider.cs

```diff
@@ -0,0 +1,124 @@
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Fhir.Synapse.Common.Configurations;
+using Microsoft.Health.Fhir.Synapse.SchemaManagement.ContainerRegistry;
+using Microsoft.Health.Fhir.Synapse.SchemaManagement.Exceptions;
+using Microsoft.Health.Fhir.TemplateManagement.ArtifactProviders;
+using Microsoft.Health.Fhir.TemplateManagement.Client;
+using Microsoft.Health.Fhir.TemplateManagement.Exceptions;
+using Microsoft.Health.Fhir.TemplateManagement.Models;
+using Microsoft.Health.Fhir.TemplateManagement.Utilities;
+using Newtonsoft.Json;
+
+namespace Microsoft.Health.Fhir.Synapse.Core.DataFilter
+{
+    public class ContainerRegistryFilterProvider : IFilterProvider
+    {
+        private readonly string _imageReference;
+        private readonly IContainerRegistryTokenProvider _containerRegistryTokenProvider;
+        private readonly ILogger<ContainerRegistryFilterProvider> _logger;
+        private readonly string _configName = "filterConfiguration.json";
+
+        public ContainerRegistryFilterProvider(
+            IOptions<FilterLocation> filterLocation,
+            IContainerRegistryTokenProvider containerRegistryTokenProvider,
+            ILogger<ContainerRegistryFilterProvider> logger)
+        {
+            EnsureArg.IsNotNull(filterLocation, nameof(filterLocation));
+            _containerRegistryTokenProvider = EnsureArg.IsNotNull(containerRegistryTokenProvider, nameof(containerRegistryTokenProvider));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _imageReference = filterLocation.Value.FilterImageReference;
+            _configName = filterLocation.Value.FilterConfigurationFileName;
+        }
+
+        public async Task<FilterConfiguration> GetFilterAsync(CancellationToken cancellationToken)
+        {
+            ImageInfo imageInfo;
+            try
+            {
+                imageInfo = ImageInfo.CreateFromImageReference(_imageReference);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(string.Format("Failed to parse the schema image reference {0} to image information. Reason: {1}.", _imageReference, ex.Message));
+                throw new ContainerRegistryFilterException(string.Format("Failed to parse the schema image reference {0} to image information. Reason: {1}.", _imageReference, ex.Message), ex);
+            }
+
+            var accessToken = await _containerRegistryTokenProvider.GetTokenAsync(imageInfo.Registry, cancellationToken);
+
+            try
+            {
+                AcrClient client = new AcrClient(imageInfo.Registry, accessToken);
+                var filterConfigurationProvider = new OciArtifactProvider(imageInfo, client);
+                var acrImage = await filterConfigurationProvider.GetOciArtifactAsync(cancellationToken);
```

Comment:
Do we need retry here?

Reply:
If it fails, will the job fail and retried in job hosting service?

Reply:
we need to add retry for any possible transient failure. Job level retry should only for unexpected error.

Reply:
After offline discussion, we can add retry in another task together with ACR for customized schema.